addons/l10n_es_edi_tbai/models/web_services.py

```diff
@@ -0,0 +1,245 @@
+        return self._get_url('sigpolicy_digest')
+
+    def get_xsd_urls(self):
+        """
+        Araba and Gipuzkoa each have a single URL pointing to a zip file (which may contain more than those two XSDs)
+        Bizkaia has two URLs (one for each XSD): in that case a tuple of strings is returned (instead of a single string)
+        """
+        return self._get_url('xsd')
+
+    def get_emission_url(self, company):
+        return self._get_url(key='invoice_', company=company)
+
+    def get_cancellation_url(self, company):
+        return self._get_url(key='cancel_', company=company)
+
+    def get_qr_base_url(self, company):
+        return self._get_url(key='qr_', company=company)
+
+class ArabaWebServices(TicketBaiWebServices):
+
+    URLS = {
+        'sigpolicy': 'https://ticketbai.araba.eus/tbai/sinadura/',
+        'sigpolicy_digest': 'd69VEBc4ED4QbwnDtCA2JESgJiw+rwzfutcaSl5gYvM=',
+        'xsd': 'https://web.araba.eus/documents/105044/5608600/TicketBai12+%282%29.zip',
+        'invoice_test': 'https://pruebas-ticketbai.araba.eus/TicketBAI/v1/facturas/',
+        'invoice_prod': 'https://ticketbai.araba.eus/TicketBAI/v1/facturas/',
+        'qr_test': 'https://pruebas-ticketbai.araba.eus/tbai/qrtbai/',
+        'qr_prod': 'https://ticketbai.araba.eus/tbai/qrtbai/',
+        'cancel_test': 'https://pruebas-ticketbai.araba.eus/TicketBAI/v1/anulaciones/',
+        'cancel_prod': 'https://ticketbai.araba.eus/TicketBAI/v1/anulaciones/',
+    }
+
+    def _prepare_post_params(self, invoice, invoice_xml, cancel=False):
+        company = invoice.company_id
+        return {
+            'url': self.get_cancellation_url(company) if cancel else self.get_emission_url(company),
+            'headers': {"Content-Type": "application/xml; charset=utf-8"},
+            'pkcs12_data': company.l10n_es_tbai_certificate_id,
+            'data': etree.tostring(invoice_xml, encoding='UTF-8'),
+        }
+
+    def _process_post_response(self, response):
+        response_data = response.content.decode(response.encoding)
+        try:
+            response_xml = etree.fromstring(bytes(response_data, 'utf-8'))
+        except etree.XMLSyntaxError as e:
+            return False, e, None
+
+        # Error management
+        message = ''
+        already_received = False
+        # Get message in basque if env is in basque
+        msg_node_name = 'Azalpena' if get_lang(self.env).code == 'eu_ES' else 'Descripcion'
+        for xml_res_node in response_xml.findall(r'.//ResultadosValidacion'):
+            message_code = xml_res_node.find('Codigo').text
+            message += message_code + ": " + xml_res_node.find(msg_node_name).text + "\n"
+            if message_code in ('005', '019'):
+                already_received = True  # error codes 5/19 mean XML was already received with that sequence
+        response_code = int(response_xml.find(r'.//Estado').text)
+        response_success = (response_code == 0) or already_received
```

Comment:
No, in particular after a timeout.
We could use their "consulta" request first every time we post something that already has an index... Still, we should handle this particular response anyways because otherwise we'll confuse it for an error and allow other invoices to be posted thus breaking the chain.

Reply:
I mean that it might be good e.g. to have a message in the chatter about it e.g. and the calling method will not know in which case it was based on the return value.

Reply:
There's already a message coming from the parsed response. Not sure why the calling method would need to know about this specific case, as far as it's concerned it's a success since the edi_state can be set to posted and the document can be saved (although, once more, ideally we would use the "consulta" request to retrieve the exact document the govt has, which could in theory be slightly different from what we have).